### PR TITLE
Utils random

### PR DESCRIPTION
**Pull request by Lucia-Fonseca**
## Description
Utility function to sample galaxies from the Schechter function.
## Checklist
- [x] Follow the [Contributor Guidelines](https://github.com/skypyproject/skypy/blob/develop/CONTRIBUTING.md)
- [x] Write unit tests
- [x] Write documentation strings
- [x] Assign someone from your working team to review this pull request
- [x] Assign someone from the infrastructure team to review this pull request
## Additional
- ~It lacks unit tests~ Suggestions for more tests.
- ~Documentation should be improved.~
- Needs #issues to be linked to and a better description

### CODE REVIEW

File: skypy/utils/random.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import numpy as np
+import skypy.utils.special as special
+
+
+def schechter(alpha, x_min, x_max, resolution=100, size=None):
+    """Galaxy sampling from the Schechter function.
```
Comment (rrjbca) [inline skypy/utils/random.py:6]:
```suggestion
    """Sample from the Schechter function.
```

File: skypy/utils/random.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    Parameters
+    ----------
+    alpha : float or int
+        The alpha parameter in the Schechter function in [1]_.
+    x_min, x_max : float or int
```
Comment (rrjbca) [inline skypy/utils/random.py:12]:
```suggestion
    x_min, x_max : array_like
```

File: skypy/utils/random.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        Lower and upper bounds for the random variable x.
+    resolution : int
+        Resolution of the inverse transform sampling spline. Default is 100.
+    size: int, optional
+         Output shape of luminosity samples.  Default is None.
+
+    Returns
+    -------
+    x_sample : array_like
+        Drawn sampling from the Schechter function.
```
Comment (rrjbca) [inline skypy/utils/random.py:22]:
```suggestion
        Samples drawn from the Schechter function.
```

File: skypy/utils/random.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+    References
+    ----------
+    .. [1] Herbel J., Kacprzak T., Amara A. et al., 2017, Journal of Cosmology
+           and Astroparticle Physics, Issue 08, article id. 035 (2017)
```
Comment (rrjbca) [inline skypy/utils/random.py:27]:
```suggestion
    .. [1] https://en.wikipedia.org/wiki/Luminosity_function_(astronomy)
```

File: skypy/utils/random.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    cdf = _cdf(x, np.min(x_min), np.max(x_max), alpha)
+    t_lower = np.interp(x_min, x, cdf)
+    t_upper = np.interp(x_max, x, cdf)
+    u = np.random.uniform(t_lower, t_upper, size=size)
+    x_sample = np.interp(u, cdf, x)
+
+    return x_sample
+
+
+def _cdf(x, x_min, x_max, alpha):
```
Comment (rrjbca) [inline skypy/utils/random.py:41]:
```suggestion
def _schechter_cdf(x, x_min, x_max, alpha):
```

File: skypy/utils/random.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+    References
+    ----------
+    .. [1] Herbel J., Kacprzak T., Amara A. et al., 2017, Journal of Cosmology
+           and Astroparticle Physics, Issue 08, article id. 035 (2017)
+
+    """
+    x = np.logspace(np.log10(np.min(x_min)), np.log10(np.max(x_max)),
+                    resolution)
+    cdf = _cdf(x, np.min(x_min), np.max(x_max), alpha)
```
Comment (rrjbca) [inline skypy/utils/random.py:32]:
```suggestion
    cdf = _schechter_cdf(x, np.min(x_min), np.max(x_max), alpha)
```

File: skypy/utils/random.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    Parameters
+    ----------
+    alpha : float or int
+        The alpha parameter in the Schechter function in [1]_.
+    x_min, x_max : float or int
+        Lower and upper bounds for the random variable x.
+    resolution : int
+        Resolution of the inverse transform sampling spline. Default is 100.
+    size: int, optional
+         Output shape of luminosity samples.  Default is None.
```
Comment (rrjbca) [inline skypy/utils/random.py:17]:
```suggestion
        Output shape of samples. Default is None.
```

**Comment by ntessore** [comment]
Just a FYI this is already being done in #79 as well. That PR is being held up by me trying to work out a better way of sampling than the inverse transform method.

**Comment by Lucia-Fonseca** [comment]
> Just a FYI this is already being done in #79 as well. That PR is being held up by me trying to work out a better way of sampling than the inverse transform method.

Oh wow! Sorry I did not mean to step on your work. Richard and I had an off-line conversation earlier today. This is meant to have a press_schechter mass function by the end of today and to update `luminosity.py`. Shall we talk off-line?

**Comment by rrjbca** [comment]
> Just a FYI this is already being done in #79 as well. That PR is being held up by me trying to work out a better way of sampling than the inverse transform method.

Sure, here we're just trying to get something done quick so we can have Press-Schechter halo mass sampling for v0.1. I imagines #79 will be v0.2 but will probably also supersede this.

File: skypy/galaxy/luminosity.py [hunk 8]
```diff
@@ -110,23 +110,10 @@ def herbel_luminosities(redshift, alpha, a_m, b_m, size=None,
...
-
-
-def _cdf(q, q_min, q_max, alpha):
-    a = special.upper_incomplete_gamma(alpha+1, q_min)
-    b = special.upper_incomplete_gamma(alpha+1, q)
-    c = special.upper_incomplete_gamma(alpha+1, q_min)
-    d = special.upper_incomplete_gamma(alpha+1, q_max)
-    return (a-b)/(c-d)
+
+    x_sample = schechter(alpha, x_min, x_max, resolution=100, size=None)
```
Comment (rrjbca) [inline skypy/galaxy/luminosity.py:114]: This should fix the unit test failing:
```suggestion
    x_sample = schechter(alpha, x_min, x_max, resolution=100, size=size)
```

File: skypy/utils/tests/test_random.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
+import numpy as np
+
+import skypy.utils.random as random
+
+
+def test_schechter_cdf():
+    # Test that sampling corresponds to sampling from the right pdf.
```
Comment (rrjbca) [inline skypy/utils/tests/test_random.py:7]:
```suggestion
    # Test the _schechter_cdf function against precomputed values
```

File: skypy/galaxy/luminosity.py [hunk 10]
```diff
@@ -54,47 +49,32 @@ def herbel_luminosities(redshift, alpha, a_m, b_m, size=None,
...
 
     Examples
     --------
     >>> import skypy.galaxy.luminosity as lum
-
-    Sample 100 luminosity values at redshift z = 1.0 with
-    a_m = -0.9408582, b_m = -20.40492365, alpha = -1.3.
-
+    >>> # Sample 100 luminosity values at redshift z = 1.0 with
+    >>> # a_m = -0.9408582, b_m = -20.40492365, alpha = -1.3.
```
Comment (rrjbca) [inline skypy/galaxy/luminosity.py:73]: The original formatting including the blank lines was correct. We want the rendered documentation to be a mix of text and command line. Please undo this change.

```suggestion

    Sample 100 luminosity values at redshift z = 1.0 with
    a_m = -0.9408582, b_m = -20.40492365, alpha = -1.3.

```

File: skypy/galaxy/luminosity.py [hunk 11]
```diff
@@ -54,47 +49,32 @@ def herbel_luminosities(redshift, alpha, a_m, b_m, size=None,
...
     >>> luminosities = lum.herbel_luminosities(1.0, -1.3, -0.9408582,
     ...                                         -20.40492365, size=100)
-
-    Sample a luminosity value for every redshift in an array z with
-    a_m = -0.9408582, b_m = -20.40492365, alpha = -1.3.
-
+    >>> # Sample a luminosity value for every redshift in an array z with
+    >>> # a_m = -0.9408582, b_m = -20.40492365, alpha = -1.3.
```
Comment (rrjbca) [inline skypy/galaxy/luminosity.py:77]: Same again here please revert this change

```suggestion

    Sample a luminosity value for every redshift in an array z with
    a_m = -0.9408582, b_m = -20.40492365, alpha = -1.3.

```
